src/bit-fields/src/lib.rs

```diff
@@ -0,0 +1,1589 @@
+    Overflow,
+    /// Given value is more than maximum value storable in bit range.
+    #[error("Given value is more than maximum value storable in bit range.")]
+    OutOfRange,
+}
+/// Error type for [`BitRange<u8, _, _>::checked_sub_assign`], [`BitRange<u16, _,
+/// _>::checked_sub_assign`], etc.
+#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
+pub enum CheckedSubAssignError {
+    /// Operation would result in underflow of bit range.
+    #[error("Operation would result in underflow of bit range.")]
+    Underflow,
+    /// Given value is more than maximum value storable in bit range.
+    #[error("Given value is more than maximum value storable in bit range.")]
+    OutOfRange,
+}
+/// Error type for [`BitRange<u8, _, _>::checked_assign`], [`BitRange<u16, _, _>::checked_assign`],
+/// etc.
+// There will only ever be 1 possible error as a result of `CheckedAssign`, thus we will never alter
+// the layout of this struct.
+#[allow(clippy::exhaustive_structs)]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CheckedAssignError;
+impl std::error::Error for CheckedAssignError {}
+impl fmt::Display for CheckedAssignError {
+    /// ```
+    /// assert_eq!(
+    ///     bit_fields::CheckedAssignError.to_string(),
+    ///     "Given value is greater than maximum storable value in bit range"
+    /// );
+    /// ```
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Given value is greater than maximum storable value in bit range"
+        )
+    }
+}
+
+// impl BitRange<..>
+macro_rules! bit_range_impl {
+    ($x:ty,$mask:ident,$max:ident) => {
+        impl<const START: u8, const END: u8> BitRange<$x, START, END> {
+            pub const MASK: $x = $mask::<START, END>();
+            /// The maximum value this range can store
+            pub const MAX: $x = $max::<START, END>();
+
+            /// Returns the value of the bit range.
+            #[must_use]
+            #[inline]
+            pub fn read(&self) -> $x {
+                <$x>::from(self)
+            }
+
+            /// Alias for [`Self::checked_assign`].
+            ///
+            /// # Errors
+            ///
+            /// When `x` is greater than the maximum storable value in `self`.
+            // `x <= Self::MAX` guarantees `x << START` is safe.
+            #[inline]
+            pub fn write(&mut self, x: $x) -> Result<(), CheckedAssignError> {
+                self.checked_assign(x)
+            }
+
+            /// Returns a pointer to underlying data.
+            #[allow(clippy::trivially_copy_pass_by_ref, clippy::as_conversions)]
+            #[must_use]
+            #[inline]
+            pub fn data(&self) -> *const $x {
+                let a = self as *const Self;
+                let b = a.cast::<$x>();
+                // SAFETY:
+                // This is safe when `self` is within a bit field.
+                unsafe { b.sub(1) }
+            }
```

Comment:
Shouldn't this entire method be marked unsafe, as otherwise it would be possible to trigger undefined behavior in safe Rust code? 

According to the documentation of [ptr::sub](https://doc.rust-lang.org/std/primitive.pointer.html#method.sub) "Both the starting and resulting pointer must be either in bounds or one byte past the end of the same [allocated object](https://doc.rust-lang.org/std/ptr/index.html#allocated-object)", which means that if `self` is not within a bitfield (for example if its the first field in a `#[repr(C)]`struct), then this method triggers undefined behavior even if the resulting pointer is not dereferenced in unsafe code later on. 

Thus the "self needs to be part of a bit field" should be the safety invariant of the entire method. Same thing for all other methods doing pointer arithmetic in this way.

Reply:
Yes, I think you are right. I will change this.

I was considering an approach where the bits and bit ranges had lifetimes such that they couldn't be used once their bit field is dropped, I haven't been able to implement this yet. If you have any insight here I think this would be a better safer approach.